semantic/poly_types.go

```diff
@@ -822,6 +837,40 @@ func (k ObjectKind) resolvePolyType(kinds map[Tvar]Kind) (PolyType, error) {
 	return NewObjectPolyType(properties, k.lower, k.upper), nil
 }
 
+// NullableKind indicates that it is possible for this
+// variable to be the null type if no other type is
+// more appropriate.
+type NullableKind struct {
+	T PolyType
+}
+
+func (NullableKind) MonoType() (Type, bool)        { return Nil, true }
+func (NullableKind) freeVars(*Constraints) TvarSet { return nil }
+func (n NullableKind) resolveType(kinds map[Tvar]Kind) (Type, error) {
+	return Nil, nil
+}
+func (n NullableKind) resolvePolyType(kinds map[Tvar]Kind) (PolyType, error) {
+	return n.T, nil
```

Comment:
Why is this not `return n.T.resolvePolyType(kinds)`?

Reply:
It causes an infinite loop is mostly the reason. But the idea behind it is that when you are consulting `NullableKind`, you have an unresolved `PolyType` otherwise you wouldn't be consulting the `Kind` to determine what the type is.

Reply:
Yes you are definitely correct. Resolving this conversation.